catalog/app/components/Preview/quick/Markdown/Skeleton.tsx

```diff
@@ -0,0 +1,23 @@
+import * as React from 'react'
+import * as M from '@material-ui/core'
+
+import Skel from 'components/Skeleton'
+
+const useSkeletonStyles = M.makeStyles((t) => ({
+  line: {
+    height: t.spacing(3),
+    marginBottom: t.spacing(1),
+  },
+}))
+
+export default function Skeleton() {
+  const classes = useSkeletonStyles()
+  const lines = [80, 50, 100, 60, 30, 80, 50, 100, 60, 30, 20, 70]
```

Comment:
you can move this out of component to avoid allocating/computing this on every render

Reply:
Ok, but probably it makes sense to compute it on render only (only when the quick preview is toggled), otherwise it is computed on every Bucket page even without Markdown.

Reply:
if you move it out of the component, it will be allocated/computed once during the first time module is imported. that's far better that doing it on every render (tho in this particular case the perf impact is negligible ofc)

Reply:
I'm ok with both variants. But I don't think one way is far better than the other, because on the one hand we have array on every Bucket page, and on another only for those who edit markdown files and preview them with "Preview" switch. And skeleton renders for a short amount of time when the js module is loaded.

Reply:
Yes, on every bucket page, loaded once. It doesn't recompute on navigation inside SPA.
So, if we have 1000 customers that open bucket page once, we have 1000 invocations.
How much of these 1000 customers will ever use "Preview", I don't know.

If that wasn't a tiny array, but some heavy computation, costing 0.5 sec to process, would it be better to load every time Bucket page is loaded, or delay Preview function once for 0.5 sec?

Reply:
i think you're missing the point, it's not about delaying the Preview or Bucket components, it's about unnecessarily executing the same computation over and over.
especially if it was a heavy computation, you should've put that outside of the render function (or inside a useMemo at least), bc it runs on every render, and you don't control how many times react calls that render.
so, if we really cared about perf here, we should've made that initialization **lazy** and **memoized**, so that it's computed once and only on-demand

Reply:
There are unnecessary executions in both variants. In the first variant, execution happens for everyone opening the page, the second is somewhat heavier but happens only in rare cases and interim. Why is first variant is superior?

```
// contributes to everyone

const once = [1, 2, 3]

function RareCase1() {
  return once
}

---

// contributes more heavily but only in rare cases

function RareCase2() {
  const everytime = [1, 2, 3]
  return everytime
}
```

Reply:
you allocate a constant once (deterministically) and then you use it further in the code (or you don't, but that doesn't matter that much, bc the perf impact of doing that during module loading/init is negligible in the grand scheme of things, given there are other things, e.g. functions in that module, that are being initialized as well).

in the second approach, you allocate that array again and again on every render.
i repeat: it doesn't matter much in this particular case, but it would be very undesirable for heavier computations

Reply:
You’re right that React can re-render a component any number of times, especially if it has state or props. In this case, the component doesn't have props and includes state via the Material-UI `useStyles` hook. While these hooks are usually memoized, re-renders are still technically possible.

That said, my point was based on practical observations. In this specific usage, the component renders only once. I didn’t mean it will always behave this way, just that it does in this context.

With that in mind, and for clarification, when you say “agree on all the other points,” does that include the point that putting the array outside the component is not inherently 100% better than keeping it inside? Specifically, considering this component’s current usage and characteristics, both approaches have trade-offs, and one is not definitively superior.